CadbaseLibrary/CadbaseLibrary.FCMacro

```diff
@@ -0,0 +1,837 @@
+                        fns.append(modification_path + "/" + file_of_fileset.filename)
+
+                    inputs = zip(urls, fns)
+                    download_parallel(inputs)
+                    FreeCAD.Console.PrintWarning('Download is completed (')
+                    FreeCAD.Console.PrintWarning(len(urls))
+                    FreeCAD.Console.PrintWarning(' file(s)\n')
+
+    FreeCAD.Console.PrintWarning('Finished\n')
+
+def download_parallel(args):
+    # FreeCAD.Console.PrintWarning('Start download data\n')
+
+    t0 = time.time()
+
+    results = ThreadPool(cpu_count() - 1).imap_unordered(get_file, args)
+
+    for result in results:
+        FreeCAD.Console.PrintWarning('\npath: ')
+        FreeCAD.Console.PrintWarning(result[0])
+        FreeCAD.Console.PrintWarning(' time (s): ')
+        FreeCAD.Console.PrintWarning(result[1])
+
+    FreeCAD.Console.PrintWarning('\nTotal time:')
+    FreeCAD.Console.PrintWarning(time.time() - t0)
+    FreeCAD.Console.PrintWarning('\n')
+
+def get_file(args):
+    t0 = time.time()
+    url, filepath = args[0], args[1]
+    manager = QtNetwork.QNetworkAccessManager()
+
+    try:
+        request = QtNetwork.QNetworkRequest(QtCore.QUrl(url))
+        # request.setRawHeader(b"Content-Type", b"application/json");
+        request.setRawHeader(b'User-Agent', user_agent)
+
+        reply = manager.get(request)
+        loop = QtCore.QEventLoop()
+        reply.finished.connect(loop.quit)
+        loop.exec_()
+
+        if reply.error() == QtNetwork.QNetworkReply.NoError:
+
+            responseData = reply.readAll()
+
+            with open(filepath, 'wb') as f:
+                f.write(responseData)
+
+            return (filepath, time.time() - t0)
+        else:
+            FreeCAD.Console.PrintWarning('Error')
+    except Exception as e:
+        FreeCAD.Console.PrintWarning('Exception in download file: ')
+        FreeCAD.Console.PrintWarning(e)
+
+class CDBSApi:
+
+    def __init__(self, query, LIBRARYPATH):
+
+        FreeCAD.Console.PrintWarning('Getting data...\n')
+        self.doRequest(query, LIBRARYPATH)
+
+
+    def doRequest(self, query, LIBRARYPATH):
+
+        api_url = param.GetString('api-url', '')
+        req = QtNetwork.QNetworkRequest(QtCore.QUrl(api_url))
+
+        auth_header = 'Bearer ' + param.GetString('auth-token', '')
+        header = {'Authorization': auth_header}
+
+        req.setRawHeader(b"Content-Type", content_type);
+        req.setRawHeader(b"Authorization", auth_header.encode());
+
+        body = json.dumps(query).encode('utf-8')
+
+        self.nam = QtNetwork.QNetworkAccessManager()
+        reply = self.nam.post(req, body)
+        loop = QtCore.QEventLoop()
+        reply.finished.connect(loop.quit)
+        loop.exec_()
+        self.handleResponse(reply, LIBRARYPATH)
+
+    def handleResponse(self, reply, LIBRARYPATH):
+
+        new_info_file = LIBRARYPATH + "/" + RESPONSEFILE
+
+        if os.path.exists(new_info_file):
```

Comment:
I already saw this somewhere, please put these lines into a function.

Reply:
Corrected, a new function named `clearPath`.